tasktiger/worker.py

```diff
@@ -149,30 +149,66 @@ def _worker_queue_expired_tasks(self):
+        # this task (i.e. we don't release the lock unless we've exhausted our
+        # batch size, which will hopefully never happen)
+        lock = Lock(self.connection, self._key('lock', 'queue_expired_tasks'),
+                    timeout=self.config['REQUEUE_EXPIRED_TASKS_INTERVAL'])
+
+        acquired = lock.acquire(blocking=False)
+        if acquired:
+            now = time.time()
+
+            # Get a batch of expired tasks.
+            task_data = self.scripts.get_expired_tasks(
+                self.config['REDIS_PREFIX'],
                 now - self.config['ACTIVE_TASK_UPDATE_TIMEOUT'],
-                now,
-                if_exists=('noupdate',),
-                on_success=('update_sets', queue,
-                            self._key(ACTIVE), self._key(QUEUED)),
+                self.config['REQUEUE_EXPIRED_TASKS_BATCH_SIZE']
             )
-            # XXX: Ideally this would be atomic with the operation above.
-            if result:
-                self.log.info('queueing expired tasks', task_ids=result)
-                self.connection.publish(self._key('activity'), queue)
+
+            for (queue, task_id) in task_data:
+                try:
+                    task = Task.from_id(self.tiger, queue, ACTIVE, task_id)
+                    if task.should_retry_on(JobTimeoutException):
+                        self.log.info('queueing expired task',
+                                      queue=queue, task_id=task_id)
+
+                        # Task is idempotent and can be requeued. If the task
+                        # already exists in the QUEUED queue, don't change its
+                        # time.
+                        task._move(from_state=ACTIVE,
+                                   to_state=QUEUED,
+                                   when=now,
+                                   mode='nx')
+                    else:
+                        self.log.error('failing expired task',
+                                       queue=queue, task_id=task_id)
+
+                        # Assume the task can't be retried and move it to the error
+                        # queue.
+                        task._move(from_state=ACTIVE, to_state=ERROR, when=now)
+                except TaskNotFound:
+                    # Either the task was requeued by another worker, or we
+                    # have a task without a task object.
+
+                    # XXX: Ideally, the following block should be atomic.
+                    if not self.connection.get(self._key('task', task_id)):
+                        self.log.error('not found',
+                                       queue=queue, task_id=task_id)
+                        task = Task(self.tiger, queue=queue,
+                                    _data={'id': task_id}, _state=ACTIVE)
+                        task._move()
+
+            # Keep requeueing if we've exhausted our batch size, otherwise
```

Comment:
Normally, if we don't release the lock, it expires after its timeout. As stated in the comment above, "we use the lock both to unnecessarily prevent multiple workers from requeueing expired tasks, as well as to space out this task (i.e. we don't release the lock unless we've exhausted our batch size, which will hopefully never happen)". So if there are more expired tasks left to requeue, the next worker executing `_worker_queue_expired_tasks` will be able to immediately pick up instead of encountering a blocked lock.